### PR TITLE
[WIP] [runtime] Automatic Reference Counting

### PR DESCRIPTION
**Pull request by harlanhaskins**
This PR tracks progress of the Automatic Reference Counting transition

- [ ] Spec out retain count conventions for function/value boundaries
- [x] Implement runtime structures
- [ ] Implement IRGen-time calls into runtime functions
- [ ] Implement automatic retain/release call insertions

### CODE REVIEW

File: Sources/Runtime/arc.cpp [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+#include <atomic>
+#include <mutex>
+
+using namespace trill;
+
+/**
+ A RAII class that will lock a mutex when it's constructed
+ and unlock the mutex once it's destroyed.
+ */
+struct LockRAII {
```
Comment (segiddins) [inline Sources/Runtime/arc.cpp]: pretty sure the stdlib has something like a `std::lock_guard`

File: Sources/Runtime/arc.cpp [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    return reinterpret_cast<void *>(reinterpret_cast<uintptr_t>(this) +
+                                    sizeof(RefCountBox));
+  }
+};
+
+/**
+ A convenience struct that performs the arithmetic necessary to work with a
+ \c RefCountBox.
+ */
+struct RefCounted {
```
Comment (segiddins) [inline Sources/Runtime/arc.cpp:73]: what's the advantage of the box vs using atomic incr/decr on the retain count?

File: Sources/Runtime/arc.cpp [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+    box->retainCount--;
+    if (box->retainCount == 0) {
+      dealloc();
+    }
+  }
+
+  /**
+   Deallocates the value inside a RefCountBox.
+   */
+  void dealloc() {
```
Comment (segiddins) [inline Sources/Runtime/arc.cpp]: this will also need to handle calling `deinit`

Comment (harlanhaskins) [reply Sources/Runtime/arc.cpp]: Good call!

**Comment by harlanhaskins** [comment]
As you can tell, my C++-fu is lacking 😅

**Comment by segiddins** [comment]
@harlanhaskins no worries :)
Mine is barely existent, just drawing on painful memories :P

File: Sources/Runtime/arc.cpp (hunk 3 above)
Comment (segiddins) [reply Sources/Runtime/arc.cpp]: (which will involve emitting a function pointer to the deinitializer in the type metadata, which I'm not sure if we're already doing)

Comment (harlanhaskins) [reply Sources/Runtime/arc.cpp]: This box is agnostic of metadata. It should just hold a deinitializer reference in the box alongside the retain count.

File: Sources/Runtime/arc.cpp (hunk 2 above)
Comment (harlanhaskins) [reply Sources/Runtime/arc.cpp:73]: With atomics, I can't check for `retainCount` over/underflow and keep the atomicity guarantee.

Comment (segiddins) [reply Sources/Runtime/arc.cpp:73]: Definitely would advise looking at what the ObjC RT does here -- I trust Greg way more than either of us :P

File: Sources/Runtime/arc.cpp [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+      trill_fatalError("object deallocated with retain count > 0");
+    }
+    if (box->deinit != nullptr) {
+      box->deinit(box->value());
+    }
+    free(box->value());
+  }
+};
+
+void *_Nonnull trill_allocateIndirectType(size_t size) {
```
Comment (segiddins) [inline Sources/Runtime/arc.cpp]: needs to now take the de-initializer as well

Comment (harlanhaskins) [reply Sources/Runtime/arc.cpp]: Yep, sorry I did that locally after committing.

**Comment by segiddins** [comment]
We should add an example for de-initializers

File: Sources/IRGen/ValueIRGen.swift [hunk 5]
```diff
@@ -174,6 +172,40 @@ extension IRGenerator {
...
+
+    if decl.isIndirect {
+      let irType = resolveLLVMType(canDestTy)
+      let allocaForRelease = createEntryBlockAlloca(currentFunction!.functionRef!,
+                                                    type: irType,
+                                                    name: "releasable",
+                                                    storage: .reference)
+      allocaForRelease.write(value)
+      codegenRetain(value)
+      releaseCalls.append({ [unowned self] in
```
Comment (segiddins) [inline Sources/IRGen/ValueIRGen.swift:201]: we need to only be adding a release call if its a var assign exr of local scope, I think?

Comment (harlanhaskins) [reply Sources/IRGen/ValueIRGen.swift:201]: This code is so wrong, I'm gonna close this PR.

**Pull request closed without merging**